### PR TITLE
Send window updates based on examined rather than consumed.

### PR DESCRIPTION
**Pull request by jkotalik**
Fixes: https://github.com/aspnet/AspNetCore/issues/8105

Will merge https://github.com/aspnet/AspNetCore/pull/8198 separate from this change (the other half).

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs
Comment (davidfowl) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs]: Looking at the other failure I'm handling in System.IO.Pipelines, I think this is fundamentally busted 😢 . It's impossible to use SequencePosition across read calls today.

Comment (jkotalik) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs]: Any change we can make to SequencePosition to make this work?

**Comment by jkotalik** [comment]
Need https://github.com/aspnet/AspNetCore/pull/8198 before this can be merged.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs
Comment (Tratcher) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs]: This only make sense if _totalExamined is _totalExamined per _readResult. You would need to reset _totalExamined when you update _readResult.

Comment (jkotalik) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs]: You are correct. I think I can fix this via using consumed. I'll update it shortly.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs]: Nit: Rename to OnDataExamined.

Also, when you fix the newlyExamined logic, you'll want to pass in newlyExamined instead of dataLength.

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2StreamTests.cs
Comment (halter73) [inline src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2StreamTests.cs]: ?

**Comment by jkotalik** [comment]
@halter73 @Tratcher I updated the logic. I _think_ this works. I'll add tests appropriately once we confirm my logic makes sense.

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2StreamTests.cs
Comment (jkotalik) [reply src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2StreamTests.cs]: Who knows how that happened ¯\\_(ツ)_/¯

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs]: File an issue, and make the pipe Pause/ResumeWriterThreashold long.MaxValue in the meantime.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs
Comment (Tratcher) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs]: So _totalExaminedInReadResult is the data already examined in the next read result, not this one. That's why you can subtract out consumed.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs
Comment (jkotalik) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs]: We should get updated dependencies before EOD. Anyways, did we decide what the default will be for the thresholds?

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs
Comment (Tratcher) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs]: 25 old bytes and 25 new ones?

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs
Comment (Tratcher) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs]: 45 old bytes and 5 new ones?

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs
Comment (Tratcher) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs]: The default wouldn't need to change with the new design.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs]: Wouldn't it be way easier to keep around a `_lastExamined` SequencePosition and have the following?

```
    var newlyExamined = _readResult.Buffer.Slice(_lastExamined, consumed).Length;
```

Comment (jkotalik) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs]: Resolved offline. We can't store the last examined in a field between calls.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs
Comment (jkotalik) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs]: Yep.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs
Comment (jkotalik) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs]: Yep.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs
Comment (jkotalik) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs]: I'll rename the variable.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs]: Special case when examined equals consumed, and don't double count the length from start to consumed.

**Comment by jkotalik** [comment]
@aspnet/brt @mikaelm12 latest macOS failure.
```
  com.microsoft.signalr.HubConnectionTest > TransportAllUsesLongPollingWhenServerOnlySupportLongPolling() FAILED
      java.lang.RuntimeException at HubConnectionTest.java:1245
          Caused by: java.lang.Exception at HubConnectionTest.java:1245

```

**Comment by mikaelm12** [comment]
> latest macOS failure. 

Noted. I'll file an issue and fix it. Thanks

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs]: Calculate length from consumed to examined.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs
Comment (Tratcher) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs]: Already examined in next read result?

**Comment by jkotalik** [comment]
Test failure is with InMemory process crash again. :/. 

Besides that 🆙 📅

**Comment by jkotalik** [comment]
@aspnet/brt MacOS timeout near razor component stuff.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs
Comment (davidfowl) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs]: ZOMG the performance :sad:

Comment (jkotalik) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs]: How would you fix it? :)

**Comment by anurse** [comment]
Yeah, we have some thoughts on timeouts, we'll be talking about this on Monday I think.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs
Comment (davidfowl) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs]: Measure first

Comment (davidfowl) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs]: OK we don’t need to gate on that.

**Comment by jkotalik** [comment]
Port conflict issue with ANCM OutOfProc with music store. https://dev.azure.com/dnceng/public/_build/results?buildId=121641

**Comment by jkotalik** [comment]
@halter73 @davidfowl when you get the chance, can you hook me up with an approval 😄

**Review by halter73 — APPROVED** [review]

**Pull request merged**
